File: pjecalc-fonte/negocio/br/jus/trt8/pjecalc/negocio/dominio/calculo/inss/PeriodoDoINSSComOpcaoSimples.java
/*
 * Decompiled with CFR 0.152.
 * 
 * Could not load the following classes:
 *  javax.persistence.Column
 *  javax.persistence.Entity
 *  javax.persistence.FetchType
 *  javax.persistence.GeneratedValue
 *  javax.persistence.GenerationType
 *  javax.persistence.Id
 *  javax.persistence.JoinColumn
 *  javax.persistence.ManyToOne
 *  javax.persistence.SequenceGenerator
 *  javax.persistence.Table
 *  javax.persistence.Temporal
 *  javax.persistence.TemporalType
 *  javax.persistence.Version
 *  org.jboss.seam.annotations.Name
 */
package br.jus.trt8.pjecalc.negocio.dominio.calculo.inss;

import br.jus.trt8.pjecalc.base.comum.Competencia;
import br.jus.trt8.pjecalc.base.comum.EntidadeBase;
import br.jus.trt8.pjecalc.base.comum.Periodo;
import br.jus.trt8.pjecalc.base.comum.Utils;
import br.jus.trt8.pjecalc.negocio.comum.GerenciadorDeValidadores;
import br.jus.trt8.pjecalc.negocio.comum.MensagemDeRecurso;
import br.jus.trt8.pjecalc.negocio.comum.exceptions.NegocioException;
import br.jus.trt8.pjecalc.negocio.comum.validators.GreaterOrEqualThan;
import br.jus.trt8.pjecalc.negocio.comum.validators.LimitedTo100Years;
import br.jus.trt8.pjecalc.negocio.comum.validators.Required;
import br.jus.trt8.pjecalc.negocio.constantes.Mensagens;
import br.jus.trt8.pjecalc.negocio.dominio.calculo.inss.Inss;
import br.jus.trt8.pjecalc.negocio.dominio.calculo.inss.RepositorioDePeriodoDoINSSComOpcaoSimples;
import java.util.Date;
import java.util.List;
import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;
import javax.persistence.SequenceGenerator;
import javax.persistence.Table;
import javax.persistence.Temporal;
import javax.persistence.TemporalType;
import javax.persistence.Version;
import org.jboss.seam.annotations.Name;

@Entity
@Table(name="TBINSSOPCAOSIMPLES")
@SequenceGenerator(name="SQINSSOPCAOSIMPLES", sequenceName="SQINSSOPCAOSIMPLES", allocationSize=1)
@Name(value="periodoDoINSSComOpcaoSimples")
public class PeriodoDoINSSComOpcaoSimples
extends EntidadeBase {
    private static final long serialVersionUID = -2616110638111308573L;
    @Id
    @GeneratedValue(strategy=GenerationType.SEQUENCE, generator="SQINSSOPCAOSIMPLES")
    @Column(name="IIDPERIODOOPCAOSIMPLES")
    private final Long id = null;
    @Version
    @Column(name="IIDVERSAO")
    private Long versao = 0L;
    @ManyToOne(fetch=FetchType.LAZY)
    @JoinColumn(name="IIDINSSCALCULO")
    private Inss inss;
    @Column(name="DDTINICIAL", nullable=false)
    @Temporal(value=TemporalType.DATE)
    @Required
    @LimitedTo100Years
    private Date dataInicioSimples;
    @Column(name="DDTFINAL", nullable=false)
    @Temporal(value=TemporalType.DATE)
    @Required
    @GreaterOrEqualThan(value="dataInicioSimples")
    @LimitedTo100Years
    private Date dataTerminoSimples;

    public PeriodoDoINSSComOpcaoSimples() {
        super(RepositorioDePeriodoDoINSSComOpcaoSimples.class);
    }

    @Override
    public Object obterChavePrimaria() {
        return this.getId();
    }

    @Override
    public Long getVersao() {
        return this.versao;
    }

    @Override
    public void setVersao(Long versao) {
        this.versao = versao;
    }

    public Long getId() {
        return this.id;
    }

    public Inss getInss() {
        return this.inss;
    }

    public void setInss(Inss inss) {
        this.inss = inss;
    }

    public Date getDataInicioSimples() {
        return this.dataInicioSimples;
    }

    public void setDataInicioSimples(Date dataInicioSimples) {
        this.dataInicioSimples = dataInicioSimples;
    }

    public Date getDataTerminoSimples() {
        return this.dataTerminoSimples;
    }

    public void setDataTerminoSimples(Date dataTerminoSimples) {
        this.dataTerminoSimples = dataTerminoSimples;
    }

    public Periodo getPeriodo() {
        if (Utils.naoNulo(this.dataInicioSimples) && Utils.naoNulo(this.dataTerminoSimples)) {
            return new Periodo(this.dataInicioSimples, this.dataTerminoSimples);
        }
        return null;
    }

    public boolean isPeriodoCoincidenteCom(PeriodoDoINSSComOpcaoSimples periodo) {
        return this.getPeriodo().isDatasCoincidentesCom(periodo.getPeriodo());
    }

    @Override
    public PeriodoDoINSSComOpcaoSimples validar() {
        GerenciadorDeValidadores.getInstance().validar(PeriodoDoINSSComOpcaoSimples.class, this);
        NegocioException excecao = new NegocioException();
        if (Utils.naoNulos(this.dataInicioSimples, this.dataTerminoSimples)) {
            Date limite;
            Competencia inicial = Competencia.getInstance(this.dataInicioSimples);
            Competencia termino = Competencia.getInstance(this.dataTerminoSimples);
            Date date = limite = Utils.naoNulo(this.getInss().getCalculo().getDataDemissao()) ? this.getInss().getCalculo().getDataDemissao() : this.getInss().getCalculo().getDataTerminoCalculo();
            if (inicial.isAnteriorA(this.getInss().getCalculo().getDataAdmissao()) || inicial.isApos(limite)) {
                excecao.adicionarMensagemDeRecurso(new MensagemDeRecurso("dataInicioSimples", Mensagens.MSG0004, "In\u00edcio"));
            }
            if (termino.isAnteriorA(this.getInss().getCalculo().getDataAdmissao()) || termino.isApos(limite)) {
                excecao.adicionarMensagemDeRecurso(new MensagemDeRecurso("dataTerminoSimples", Mensagens.MSG0004, "Fim"));
            }
        }
        if (!excecao.getMensagensDeRecurso().isEmpty()) {
            throw excecao;
        }
        for (PeriodoDoINSSComOpcaoSimples periodo : this.inss.getPeriodosComOpcaoSimples()) {
            if (!this.isPeriodoCoincidenteCom(periodo)) continue;
            throw new NegocioException(new MensagemDeRecurso("dataTerminoSimples", Mensagens.MSG0024, new Object[0]));
        }
        return this;
    }

    @Override
    public void salvar() {
        this.validar();
        super.salvar();
    }

    public static List<PeriodoDoINSSComOpcaoSimples> obterTodosPor(Inss inss) {
        return PeriodoDoINSSComOpcaoSimples.getRepositorio(RepositorioDePeriodoDoINSSComOpcaoSimples.class).obterTodosPor(inss);
    }
}

File: src/lib/pjecalc/core/dominio/calculo/inss/periodo-do-inss-com-opcao-simples.ts
/**
 * PJe-Calc v2.15.1 — PeriodoDoINSSComOpcaoSimples
 * Porte 1:1 de: br.jus.trt8.pjecalc.negocio.dominio.calculo.inss.PeriodoDoINSSComOpcaoSimples
 *
 * Ref Java: pjecalc-fonte/.../calculo/inss/PeriodoDoINSSComOpcaoSimples.java
 *
 * Representa um intervalo em que o empregador optou pelo Simples Nacional.
 * Durante o período, a alíquota empresa + SAT é zerada (recolhimento unificado
 * no Simples). Configurado por Inss.adicionar(PeriodoDoINSSComOpcaoSimples).
 */
import { Periodo } from '../../../base/comum/periodo';
import { Competencia } from '../../../base/comum/competencia';
import { MensagemDeRecurso } from '../../../comum/mensagem-de-recurso';
import { Mensagens } from '../../../comum/mensagens';
import { NegocioException } from '../../../comum/exceptions/negocio-exception';
import type { Inss } from './inss';

export class PeriodoDoINSSComOpcaoSimples {
  private id: number | null = null;
  private versao: number = 0;
  private inss: Inss | null = null;
  private dataInicioSimples: Date | null = null;
  private dataTerminoSimples: Date | null = null;

  getId(): number | null { return this.id; }
  setId(id: number): void { this.id = id; }
  getVersao(): number { return this.versao; }
  setVersao(v: number): void { this.versao = v; }

  getInss(): Inss | null { return this.inss; }
  setInss(inss: Inss | null): void { this.inss = inss; }

  getDataInicioSimples(): Date | null { return this.dataInicioSimples; }
  setDataInicioSimples(d: Date | null): void { this.dataInicioSimples = d; }

  getDataTerminoSimples(): Date | null { return this.dataTerminoSimples; }
  setDataTerminoSimples(d: Date | null): void { this.dataTerminoSimples = d; }

  /** getPeriodo (Java linha 128) */
  getPeriodo(): Periodo | null {
    if (this.dataInicioSimples && this.dataTerminoSimples) {
      return new Periodo(this.dataInicioSimples, this.dataTerminoSimples);
    }
    return null;
  }

  /** isPeriodoCoincidenteCom (Java linha 135) */
  isPeriodoCoincidenteCom(outro: PeriodoDoINSSComOpcaoSimples): boolean {
    const a = this.getPeriodo();
    const b = outro.getPeriodo();
    if (!a || !b) return false;
    return a.isDatasCoincidentesCom(b);
  }

  /**
   * `validar` — porte 1-a-1 de PeriodoDoINSSComOpcaoSimples.java:140-163.
   *
   * Regras:
   *   1. Datas inicial e final devem estar entre Admissão e
   *      (Demissão ?? DataTérminoCalculo) — senão: MSG0004 em
   *      "dataInicioSimples" e/ou "dataTerminoSimples".
   *   2. Não pode coincidir com outro período Simples do mesmo Inss
   *      (MSG0024 em "dataTerminoSimples") — lança direto, sem acumular.
   *
   * Depende de `Calculo` já ter `dataAdmissao`, `dataDemissao` e
   * `dataTerminoCalculo` populados.
   */
  validar(): PeriodoDoINSSComOpcaoSimples {
    const excecao = new NegocioException();

    if (this.inss == null) return this; // sem INSS linkado, ignora
    const calculo = this.inss.getCalculo();
    if (calculo == null) return this;

    if (this.dataInicioSimples != null && this.dataTerminoSimples != null) {
      const inicial = Competencia.getInstance(this.dataInicioSimples);
      const termino = Competencia.getInstance(this.dataTerminoSimples);
      const limite =
        calculo.getDataDemissao() ?? calculo.getDataTerminoCalculo();
      const admissao = calculo.getDataAdmissao();

      if (admissao != null) {
        if (inicial.isAnteriorA(admissao) || (limite != null && inicial.isApos(limite))) {
          excecao.adicionarMensagemDeRecurso(
            new MensagemDeRecurso('dataInicioSimples', Mensagens.MSG0004, 'Início'),
          );
        }
        if (termino.isAnteriorA(admissao) || (limite != null && termino.isApos(limite))) {
          excecao.adicionarMensagemDeRecurso(
            new MensagemDeRecurso('dataTerminoSimples', Mensagens.MSG0004, 'Fim'),
          );
        }
      }
    }

    if (excecao.existeMensagensDeRecurso()) {
      throw excecao;
    }

    for (const outro of this.inss.getPeriodosComOpcaoSimples()) {
      if (outro === this) continue;
      if (this.isPeriodoCoincidenteCom(outro)) {
        throw new NegocioException(
          new MensagemDeRecurso('dataTerminoSimples', Mensagens.MSG0024),
        );
      }
    }

    return this;
  }
}
